### ISSUE TITLE
Nested data upload/download using metadata

### ISSUE THREAD

**Issue by robbarnsley** [rucio/rucio#6051]
### Description

To add the ability to upload hierarchically structured data to the datalake and retain this format after download.

### Motivation

We (SKAO) need to be able to upload data to the datalake and retain this hierarchical structure on download.

Currently the functionality in Rucio does not allow you to upload a folder that contains both files and folders as there is no native datatype that supports containerising both a file and a dataset. The nearest thing is the [recursive](https://github.com/rucio/rucio/blob/master/lib/rucio/client/uploadclient.py#L89) option in the upload client, but this has the [limitation ](https://github.com/rucio/rucio/blob/master/lib/rucio/client/uploadclient.py#L748) above.

[I have already written functionality](https://gitlab.com/ska-telescope/src/ska-rucio-extended-client/-/blob/main/src/rucio_extended_client/api/plan.py) to do this that extends upon the base client (it is rather long as it creates a "plan" first so that the operation is idempotent and can be resumed in the case of errors). This functionality uses the concepts of datasets and containers to nest the data, but in the case that a subdirectory contains both files and folders, the files are first added to a dataset with some configurable suffix (default is `__root`). This circumvents the issue discussed above.

However, during a recent test of this with a sizable chunk of data (7T, 75k total dids with 13k nested child datasets), we have encountered a fair few issues. Probably the biggest is discussed [here](https://github.com/rucio/rucio/issues/6049). In short, `conveyor-finisher` must insert a lot of rows (== # of child datasets) to a table for every DID it processes. If you consider in the above example that the number of child datasets was ~13k, even with an insert transaction taking only a couple of ms, that would be nearly half a minute per DID.

### Change

Another way of retaining hierarchical structure would be to use metadata, so my proposal is something along the lines of:

- create a "root" dataset corresponding to root folder to be uploaded & create rule
- add metadata to this root dataset that keeps nested relations, e.g. [{did1: path/on/drive/1, did2: path/on/drive/2, ...}]
- upload files and attach their DIDs to this dataset

I think this has numerous benefits over the previous method I implemented:

1. it doesn't require nesting large numbers of datasets (gets around `conveyor-finisher` issue)
2. it doesn't require a suffix to be reserved for hierarchical operations
3. could have an operation that allows single files inside the dataset to be replaced (very useful if you have a large amount of data where only a small section of it has been changed!)

So my first questions would be: Is this stupid? Am I missing something which means this wouldn't work? 

If the answer to these questions are "no", then is it possible to create a column in the `did` table for this, rather than have to use custom metadata backends? I would be quite nervous to store structural information like this in a plugin-able metadata backend.

**Comment by bari12** [comment]
It sounds feasible, but the complication here is to manage the metadata I imagine.
E.g. the set of metadata might be massive? In your example above, 75k total dids, the root dataset would have a meta entry with 75k entries?

**Comment by robbarnsley** [comment]
This is true. Presumably by recreating this hierarchical structure with containers and datasets (as I currently am) I am also creating significant numbers of rows in a table somewhere though. In terms of where this data is stored, is it "worse" if it's stored as metadata?

Perhaps it would not be so bad to store this kind of information in the custom database rather than in the `dids` table. There are certainly technologies for which these kind of numbers are not an issue (and which the likes of SKAO will need to use anyway as our general metadata requirements could be quite demanding).

**Comment by bari12** [comment]
Size is not the issue I think, I am more thinking about concurrency here. If you just `upload` a whole structure in one go, this doesn't really matter, but I am wondering if different processes update different datasets/containers (directories) in this structure, then all of them would have to update the root metadata entry, which might become a concurrency issue. Again, this might not be an issue in your workflow, but just something worth thinking about.

**Comment by robbarnsley** [comment]
Understood. So, just to try forge a path forward on this - is there any interest in adding the functionality for hierarchical uploads (using either this method or the one using native rucio data types) upstream? If so, which method is preferred?

If not, we can probably close this issue and I will continue adding functionality to our "extended" client. It would then be best to concentrate on fixing the issues in https://github.com/rucio/rucio/issues/6049 instead.

**Comment by bari12** [comment]
Thank you for raising this. We have reviewed this issue as part of our regular issue review.

We see value in this request, but we do not have the capacity to work on it within the next twelve months. 

In line with our [issue management guideline](https://rucio.cern/documentation/developer/issue_management), we are labelling it `wishlist` and closing it. This does not mean it is forgotten: wishlist items are revisited regularly during component reviews and roadmap planning, and this issue can be re-opened and added back to the pipeline as soon as interest and capacity align.
